### PR TITLE
Third level -- widely distributed combination technique

### PR DESCRIPTION
**Pull request by freifrauvonbleifrei**
This is a WIP pull request, and I am not sure whether it should ever be merged ;) however, we can discuss the changes here.

### CODE REVIEW

File: distributedcombigrid/src/sgpp/distributedcombigrid/sparsegrid/DistributedSparseGridUniform.hpp [hunk 1]
```diff
@@ -134,7 +194,16 @@ class DistributedSparseGridUniform {
 
   int commSize_;
 
-  std::vector<SubspaceSGU<FG_ELEMENT> > subspaces_;
+  std::vector<SubspaceSGU<FG_ELEMENT> > subspaces_; // subspaces of the dsg
+
+  std::vector<FG_ELEMENT> subspacesData_; // allows linear access to all subspaces data
```
Comment (freifrauvonbleifrei) [inline distributedcombigrid/src/sgpp/distributedcombigrid/sparsegrid/DistributedSparseGridUniform.hpp:199]: @mhurler123 I recall there was a reason why this additional vector was introduced, but I can't recall it. Do you remember?

Comment (mhurler123) [reply distributedcombigrid/src/sgpp/distributedcombigrid/sparsegrid/DistributedSparseGridUniform.hpp:199]: I remember that Mario copied the subspaces to an extra buffer to allow reducing them with a single MPI call. This effectively doubled the memory consumption per worker during the combination step. Note that the SubspaceSGU in my implementation only hold pointers to the subspacesData_.

See https://github.com/SGpp/DisCoTec/blob/master/distributedcombigrid/src/sgpp/distributedcombigrid/combicom/CombiCom.hpp#L776 on master:

Comment (freifrauvonbleifrei) [reply distributedcombigrid/src/sgpp/distributedcombigrid/sparsegrid/DistributedSparseGridUniform.hpp:199]: Great, thank you! I think this is the way to go for now.

**Comment by freifrauvonbleifrei** [comment]
hrm, I think that using ssh tunnels, (as with `sshCommand` in `ctparam`), we are facing the issue of TCP-over-TCP here : 
http://sites.inka.de/bigred/devel/tcp-tcp.html

In case we are really stuck with ssh, we could try to use UDP (over ssh), which might be as easy as replacing `SOCK_STREAM` with `SOCK_DGRAM` in `NetworkUtils.cpp` ...

**Comment by freifrauvonbleifrei** [comment]
through some detours across the `third-level-advection` branch, these changes are now in the `main` branch. I am closing this PR.

**Pull request closed without merging**